cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -135,6 +135,17 @@ func ValidateArgSelection(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) f
 	if len(cfg.DiscoveryTokenAPIServers) < 1 && len(cfg.DiscoveryToken) != 0 {
 		allErrs = append(allErrs, field.Required(fldPath, "DiscoveryTokenAPIServers not set"))
 	}
+
+	if len(cfg.DiscoveryFile) != 0 && len(cfg.TLSDiscoveryRootCAPubKeys) != 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, "", "TLSDiscoveryRootCAPubKeys cannot be used with DiscoveryFile"))
+	}
+
+	// TODO: convert this warning to an error after v1.8
+	if len(cfg.DiscoveryFile) == 0 && len(cfg.TLSDiscoveryRootCAPubKeys) == 0 && !cfg.TLSDiscoveryAllowUnsafe {
```

Comment:
The way it's written now, `--tls-discovery-root-ca` _only_ applies to the plain token-based case. If you're using `DiscoveryFile` with local files or a remote HTTPS endpoint, it doesn't do any validation. 

I think that's reasonable because `DiscoveryFile` should be transmitted through some secure means that protects its integrity. Either the local file will have been copied there by some secure orchestrator or the remote file will be authenticated by the web PKI and the system roots. I don't think we need to consider these cases unsafe or require the user to pass `--tls-discovery-allow-unsafe`.

I have the validation error `TLSDiscoveryRootCAPubKeys cannot be used with DiscoveryFile` to make sure you don't get a false sense of security from thinking you've pinned the CA when you're actually just relying on the security of the discovery file.